Preliminary refactoring of pathfinding

roryokane opened the pull request:
I started to implement pathfinding. I’m still working on it, but in the meantime, these refactorings can already be pulled in.

See individual commit messages for more details.

roryokane committed:
Rename Diagonal heuristic to Chebyshev

Because I find myself making a set of classes like ManhattanMovementStyle and DiagonalMovementStyle. DiagonalMovementStyle sounds like it only moves diagonally, like a bishop in chess, giving the wrong idea.

roryokane committed:
Refactor GraphFactory

All of the changes:

- convert object and class `Node` into a case class, so that `Node`s can be tested for equality
- support printing of `SLOW` nodes
- Add explanatory comments to `generateGraph`. These will hopefully make it easier to understand why SLOW is not supported, and how to implement it if slow tiles are ever actually placed on the map.
- Shorten the last section of `generateGraph` by assigning `tile` and `node` with pattern matching instead of explicitly.
- Move `generateGraph` to the top, because it’s the most important.
- Greatly simplify `inBounds` by using `forall` and `Range`.
- fix misspelling in comment

chrisbenincasa commented on src/main/scala/ayai/factories/GraphFactory.scala:
```diff
@@ -53,17 +43,26 @@ object GraphFactory {
+  }
+
+  private def inBounds(max: Int, indexes: Int*): Boolean = {
+    indexes.forall{ index => Range.inclusive(0, max).contains(index) }
```
[style] you should be able to rewrite this like:

``` scala
indexes.forall(Range.inclusive(0, max).contains)
```

roryokane replied:
Cool. Incorporated into the latest commit.

chrisbenincasa commented on src/main/scala/ayai/factories/GraphFactory.scala:
```diff
@@ -3,44 +3,34 @@ package ayai.factories
-object Node {
-  def apply(state: NodeState) = new Node(state)
+case class Node(val state: NodeState) {
+  def printableState = state match {
```
can we just override `toString` with the contents of this method instead of something completely new?

roryokane replied:
I don’t think so. `toString` isn’t just an alias for `printableState` – it adds “Node” to the string as well, e.g. “Impassable Node”. In the latest commit, I moved `printableState` to the end and made it `private` so that is clearer.

I tried changing the method to a `val` inside `toString`, but that looked more complicated.

chrisbenincasa commented on src/main/scala/ayai/components/pathfinding/DistanceHeuristic.scala:
```diff
@@ -25,13 +25,13 @@ class ManhattanDistance extends DistanceHeuristic {
  * http://en.wikipedia.org/wiki/Chebyshev_distance
  */
-class DiagonalDistance extends DistanceHeuristic {
+class ChebyshevDistance extends DistanceHeuristic {
```
:+1:

chrisbenincasa commented:
Looks good man :+1:

roryokane committed:
More GraphFactory refactoring

- simplify `inBounds` according to [Christian’s suggestion](https://github.com/ayai-ai/ayai-backend/pull/21#discussion_r28920614)
- move `toString` to the top of Node to make the necessity for `printableState` clearer, and change the string to put the adjective in front for readability
- remove an extra space
